File: Dia16/main_p2.py
#!/usr/bin/python3

# No me he sentido menos orgulloso de un programa en mi vida.
# En CPython, tarda 2 horas en ejecutarse en mi ordenador.
# Me encantaría testearlo en PyPy pero pgraph me da problemas
# y me da pereza portear este código para que use otra librería.

from pgraph import *
from collections import defaultdict
from dataclasses import dataclass

grafo = UGraph()

archivo = map(lambda x: x.rstrip('\n').split(' '), iter(open("input.txt","r").readlines()))

flujos = {}
adyacencias = {}

for linea in archivo:
	valvula = linea[1]
	flujo = int(linea[4][5:-1])
	v = grafo.add_vertex(name=valvula)
	flujos[v] = flujo
	adyacentes = [x.rstrip(',') for x in linea[9:]]
	adyacencias[v] = adyacentes

for vertex, adyacentes in adyacencias.items():
	for adyacente in adyacentes:
		v_adyacente = grafo[adyacente]
		if vertex not in v_adyacente.neighbours():
			vertex.connect(v_adyacente, cost = 1)

vertices_importantes = [v for v in grafo if flujos[v] > 0]
vertices_importantes.append(grafo['AA'])

def dijkstra(inicio: UVertex) -> dict[UVertex, int]:
	dist = defaultdict(lambda: float('inf'))
	dist[inicio] = 0
	abiertos = [x for x in grafo]
	while len(abiertos) != 0:
		u = min(abiertos, key = lambda x: dist[x])
		abiertos.remove(u)
		for v in u.neighbours():
			alt = dist[u] + 1
			if alt < dist[v]:
				dist[v] = alt
	dist = dict([(key, val) for key, val in dist.items() if key in vertices_importantes])
	return dist

distancias: dict[tuple[UVertex, UVertex], int] = defaultdict(lambda: None)

for i in range(len(vertices_importantes)):
	vertice1 = vertices_importantes[i]
	dist = dijkstra(vertice1)
	for vertice2, distancia in dist.items():
		distancias[(vertice1, vertice2)] = distancia

eliminados = set()
for v in grafo:
	if v not in vertices_importantes:
		eliminados.add(v)

if grafo['AA'] in eliminados:
	eliminados.remove(grafo['AA'])

for v in eliminados:
	grafo.remove(v)

for (v1, v2), dist in distancias.items():
	if v1 is v2:
		continue
	if v1 in v2.neighbours():
		v1.edgeto(v2).cost = dist
	else:
		v1.connect(v2, cost = dist)

arr = []
for i in range(grafo.n):
	arr1 = []
	for j in range(grafo.n):
		if grafo[i] in grafo[j].neighbours():
			arr1.append(grafo[i].edgeto(grafo[j]).cost)
		else:
			arr1.append(0)
	arr.append(arr1)
print(arr)

for i in range(grafo.n):
	if grafo['AA'] is grafo[i]:
		print(i)

# exit()
max_tiempo = 26

@dataclass
class Nodo:
	vertices: list[UVertex]
	visitados: dict[UVertex, int]
	minutos: list[int]

	def __init__(self, vert, vis, min):
		self.vertices = vert
		self.visitados = vis
		self.minutos = min

def copia(nodo: Nodo) -> Nodo:
	vert = nodo.vertices.copy()
	visi = {}
	for key, val in nodo.visitados.items():
		visi[key] = val
	min = nodo.minutos.copy()
	return Nodo(vert, visi, min)

def presion_total(nodo: Nodo) -> int:
	abiertos = nodo.visitados
	sumador = 0
	for vertice, minuto in abiertos.items():
		cantidad_flujo = flujos[vertice]*(max_tiempo - minuto)
		sumador += cantidad_flujo
	return sumador

def vecinos(nodo: Nodo) -> list[Nodo]:
	minutos = enumerate(nodo.minutos)
	minutos = sorted(minutos, key = lambda x: x[1])

	for i, minuto in minutos:
		res = []
		vertice = nodo.vertices[i]
		for vecino in vertice.neighbours():
			if vecino in nodo.visitados.keys():
				continue
			minutos_vecino = minuto + vertice.edgeto(vecino).cost
			if minutos_vecino > max_tiempo:
				continue
			nuevos_vertices = nodo.vertices.copy()
			nuevos_vertices[i] = vecino
			nuevos_minutos = nodo.minutos.copy()
			nuevos_minutos[i] = minutos_vecino
			nodo_vecino = Nodo(nuevos_vertices, nodo.visitados, nuevos_minutos)
			res.append(nodo_vecino)
		if len(res) > 0:
			return res
	return []
	
def explorar(nodo: Nodo) -> Nodo:
	for i in range(len(nodo.vertices)):
		vertice = nodo.vertices[i]
		if vertice not in nodo.visitados.keys():
			nodo_copia = copia(nodo)
			nodo_copia.minutos[i] += 1
			nodo_copia.visitados[vertice] = nodo_copia.minutos[i]
			nodo = nodo_copia
	
	vec = vecinos(nodo)
	if len(vec) == 0:
		return nodo
	max_act = (None, -1)
	for vecino in vec:
		res = explorar(vecino)
		presion = presion_total(res)
		if max_act[1] <= presion:
			max_act = (res, presion)
	return max_act[0]

inicial = Nodo([grafo['AA'], grafo['AA']], {grafo['AA']: 0}, [0, 0])
print(presion_total(explorar(inicial)))

File: Dia16/main_p2.rs
// Esto es un intento de la parte 2 en Rust.
// Los COSTES y FLUJOS son extraídos de la
// primera parte de la versión de Python de
// este programa.
// Me da un error de stack y no tengo la
// suficiente paciencia para arreglarlo.

const NODOS: usize = 16;

const COSTES: [[usize; NODOS]; NODOS] = [[0, 3, 3, 7, 8, 11, 3, 3, 5, 10, 5, 2, 5, 5, 8, 12],
								[3, 0, 5, 7, 9, 12, 2, 2, 2, 11, 7, 3, 6, 5, 10, 13],
								[3, 5, 0, 4, 6, 9, 3, 6, 7, 8, 2, 3, 6, 2, 5, 10],
								[7, 7, 4, 0, 2, 5, 5, 9, 9, 4, 5, 7, 5, 2, 8, 6],
								[8, 9, 6, 2, 0, 3, 7, 11, 11, 2, 7, 6, 3, 4, 10, 4],
								[11, 12, 9, 5, 3, 0, 10, 14, 14, 3, 10, 9, 6, 7, 13, 2],
								[3, 2, 3, 5, 7, 10, 0, 4, 4, 9, 5, 2, 5, 3, 8, 11],
								[3, 2, 6, 9, 11, 14, 4, 0, 3, 13, 8, 5, 8, 7, 11, 15],
								[5, 2, 7, 9, 11, 14, 4, 3, 0, 13, 9, 5, 8, 7, 12, 15],
								[10, 11, 8, 4, 2, 3, 9, 13, 13, 0, 9, 8, 5, 6, 12, 2],
								[5, 7, 2, 5, 7, 10, 5, 8, 9, 9, 0, 5, 8, 3, 3, 11],
								[2, 3, 3, 7, 6, 9, 2, 5, 5, 8, 5, 0, 3, 5, 8, 10],
								[5, 6, 6, 5, 3, 6, 5, 8, 8, 5, 8, 3, 0, 7, 11, 7],
								[5, 5, 2, 2, 4, 7, 3, 7, 7, 6, 3, 5, 7, 0, 6, 8],
								[8, 10, 5, 8, 10, 13, 8, 11, 12, 12, 3, 8, 11, 6, 0, 14],
								[12, 13, 10, 6, 4, 2, 11, 15, 15, 2, 11, 10, 7, 8, 14, 0]];

const FLUJOS: [usize; NODOS] = [7, 14, 6, 12, 21, 23, 0, 20, 15, 22, 19, 3, 8, 13, 17, 24];

const MAX_TIEMPO: usize = 10;

#[derive(Clone, Copy)]
struct Posicion {
	vertice: usize,
	minuto: usize
}

#[derive(Clone, Copy)]
struct Nodo {
	entidades: [Posicion; 2],
	visitados: [Option<usize>; FLUJOS.len()]
}

fn presion_total(nodo: Nodo) -> isize {
	let mut sumador = 0;
	for i in 0..nodo.visitados.len() {
		if let Some(minuto) = nodo.visitados[i] {
			sumador += (MAX_TIEMPO as isize - minuto as isize)*(FLUJOS[i] as isize);
		}
	}
	return sumador;
}

fn vecinos(nodo: Nodo) -> Vec<Nodo> {
	if nodo.entidades[0].minuto < nodo.entidades[1].minuto {
		let res = vecinos_entidad(nodo, 0);
		if res.len() != 0 {
			return res;
		}
		return vecinos_entidad(nodo, 1);
	} else {
		let res = vecinos_entidad(nodo, 1);
		if res.len() != 0 {
			return res;
		}
		return vecinos_entidad(nodo, 0);
	}
}

fn vecinos_entidad(nodo: Nodo, entidad: usize) -> Vec<Nodo> {
	let mut res = vec![];
	for i in 0..NODOS {
		if i == entidad {
			continue;
		}
		if let Some(_) = nodo.visitados[i] {
			continue;
		}
		let vertice = nodo.entidades[entidad].vertice;
		let minuto_vecino = nodo.entidades[entidad].minuto + COSTES[vertice][i];
		if minuto_vecino > MAX_TIEMPO {
			continue;
		}
		let mut entidades_vecino = nodo.entidades;
		entidades_vecino[entidad].vertice = i;
		entidades_vecino[entidad].minuto = minuto_vecino;
		let nodo_vecino = Nodo {
			entidades: entidades_vecino,
			visitados: nodo.visitados,
		};
		res.push(nodo_vecino);
	}
	res
}

fn main() {
	let ini = 6;
	let inicial = Nodo {
		entidades: [Posicion {
			vertice: ini,
			minuto: 0
		}; 2],
		visitados: [None; NODOS],
	};
	let res = presion_total(explorar(inicial));
	println!("{res}");
}

fn explorar(nodo_i: Nodo) -> Nodo {
	let mut nodo = nodo_i;
	for i in 0..nodo.entidades.len() {
		let vertice = nodo.entidades[i].vertice;
		if let Some(_) = nodo.visitados[vertice] {
			continue;
		}
		if nodo.entidades[i].minuto >= MAX_TIEMPO {
			continue;
		}
		nodo.entidades[i].minuto += 1;
		nodo.visitados[vertice] = Some(nodo.entidades[i].minuto);
	}
	
	let vecinos = vecinos(nodo);
	if vecinos.len() == 0 {
		return nodo;
	}
	let mut max_nodo = None;
	let mut max_presion = 0;
	for vecino in vecinos {
		let res = explorar(vecino);
		let presion = presion_total(res);
		if presion > max_presion {
			max_presion = presion;
			max_nodo = Some(res);
		}
	}
	if let Some(resultado) = max_nodo {
		return resultado;
	}
	panic!();
}
